### PR TITLE
adds draft for tdd-todo project for our students

### PR DESCRIPTION
**Pull request by des-des**
I know I am not following proper process here! I want to apologise for submitting this pull request without first gaining a consensus on a issue. I would be happy to move this onto a forked version of the repo, but thought I would try to get in here first as I think that would be best for all parties.

This is a very guided project, which starts with a skeleton to pull down (kinda an experiment).
Its main aims are to:
 * give the students more familiarity with the dom api
 * show students how separating business logic from side effects is essential for good unit testing

related #171, #163 

@mantagen @shiryz @JMurphyWeb right now this is pretty rough round the edges, would appreciate if you could have a look over and help tighten things up!

### CODE REVIEW

File: coursebook/fragments/projects/tdd-todo/skeleton/todo.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    generateId: (function() {
+      var idCounter = 0;
+      return function () {
+        return idCounter++;
+      }
+    })(),
+    addTodo: function (todos, newElem) {
+      // should leave the input argument todos unchanged
+      // return a new array, it should contain todos with the newElem added to the end. Do not modify the first element.
+      // add an id to the newTodo. you can use the generageId function to create an id.
```
Comment (shiryz) [inline coursebook/fragments/projects/tdd-todo/skeleton/todo.js]: Should mention newTodo before this line, or call it new array like before.
there's a typo generageId

File: coursebook/fragments/projects/tdd-todo/skeleton/todo.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    deleteTodo: function (todos, idToDelete) {
+      // should leave the input argument todos unchanged
+      // return a new array, this should not contain any todo with an id of idToDelete
+      // hint: array.filter
+      return state; //change me!
+    },
+    markTodo: function(todos, idToMark) {
+      // should leave the input argument todos unchanged
+      // return an array with the done property of idToMark toggled
```
Comment (shiryz) [inline coursebook/fragments/projects/tdd-todo/skeleton/todo.js]: What does this one mean?

File: coursebook/fragments/projects/tdd-todo/skeleton/todo.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+      // add markTodo button
+
+      // add classes for css
+
+      return todoNode;
+    },
+    render: function(state) {
+      var todoListWrapper = document.getElementById('todo-container');
+      var todoListNode = document.createElement('ul');
```
Comment (shiryz) [inline coursebook/fragments/projects/tdd-todo/skeleton/todo.js]: ul is already in the html

File: coursebook/fragments/projects/tdd-todo/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  * the first part should be done tdd, here you will create the logic needed to modify your list, and will not interact with the DOM. Use qunit to do this. create an extra file called test
+  * the second part involves rendering the todo list to the dom;
+  * the third part involves adding your own features.
+
+Some things you might want to add for the third part
+  * edit todo text
+  * style using css
+  * remove the mark button and listen for clicks on the whole todo item.
+
+Todo:
```
Comment (shiryz) [inline coursebook/fragments/projects/tdd-todo/README.md]: Add setup instructions

**Comment by JMurphyWeb** [comment]
@des-des it is more hand-holdy than I had imagined. 

Benefits:
- see good practices
- more time spent testing

Potential problems:
- They won't feel pain of not so well architected code
- They may not really appreciate what your architecture is doing for them
- Not thinking for themselves

File: coursebook/fragments/projects/tdd-todo/skeleton/todo.js (hunk 3 above)
Comment (des-des) [inline coursebook/fragments/projects/tdd-todo/skeleton/todo.js]: This gets replaced (see replaceChild below)

File: coursebook/fragments/projects/tdd-todo/skeleton/todo.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  // part 2. the Dom
+  var state = [
+    { id: -3, description: 'first todo'},
+    { id: -2, description: 'second todo'},
+    { id: -1, description: 'third todo'}
+  ]; // this is our todoList
+
+  var controller = {
```
Comment (nelsonic) [inline coursebook/fragments/projects/tdd-todo/skeleton/todo.js]: this _might_ require some explanation to people in their second week of F&C ...
But you are in the best position to gauge. 👍

**Pull request merged**